payu/envmod.py

```diff
@@ -109,4 +109,74 @@ def lib_update(required_libs, lib_name):
             return '{0}/{1}'.format(mod_name, mod_version)
 
     # If there are no libraries, return an empty string
-    return ''
\ No newline at end of file
+    return ''
+
+
+def paths_set_by_user_modules(user_modules, user_modulepaths):
+    """Search along changes PATH added by user defined modules
+    and return a set of paths added - this is used for
+    searching for the model executable"""
+    if 'MODULESHOME' not in os.environ:
+        print('payu: warning: No Environment Modules found; skipping '
+              'inspecting user module changes to PATH')
+        return set()
+
+    # Orginal environment
+    previous_env = dict(os.environ)
+    previous_modulepath = os.environ['MODULEPATH']
+
+    # Set restrict module path to only user defined module paths
+    os.environ['MODULEPATH'] = ':'.join(user_modulepaths)
+
+    # Note: Using subprocess shell to isolate changes to environment
+    paths = []
+    try:
+        # Get $PATH paths with no modules loaded
+        init_paths = paths_post_module_commands(["purge"])
+        for module in user_modules:
+            # Check if module is available
+            module_cmd = f"{os.environ['MODULESHOME']}/bin/modulecmd bash"
+            cmd = f"{module_cmd} is-avail {module}"
+            if run_cmd(cmd).returncode != 0:
+                continue
+            # TODO: Check if multiple modules are available..
+            try:
+                # Get $PATH paths post running module purge && module load
+                paths.extend(paths_post_module_commands(['purge',
+                                                         f'load {module}']))
+            except subprocess.CalledProcessError as e:
+                continue
+    finally:
+        os.environ['MODULEPATH'] = previous_modulepath
+
+    if previous_env != os.environ:
+        print(
+            "Warning: Env vars changed when inspecting paths set by modules"
+        )
+
+    # Remove inital paths and convert into a set
+    return set(paths).difference(set(init_paths))
+
+
+def paths_post_module_commands(commands):
+    """Runs subprocess module command and parse out the resulting
+    PATH environment variable"""
+    # Use modulecmd as module command is not available on compute nodes
+    module_cmds = [
+        f"eval `{os.environ['MODULESHOME']}/bin/modulecmd bash {c}`"
```

Comment:
Currently wondering about security of doing `eval` on an environment variable.

Reply:
Yeah.. Would it be better to run it in python? e.g.: https://github.com/payu-org/payu/blob/91f2831a55b61d20a0da3d5f933b6b269e0fb875/payu/envmod.py#L86-L92